### ISSUE TITLE
Show date instead of relative "X years ago" in some cases

### ISSUE THREAD

**Issue by LukasKalbertodt** [elan-ev/tobira#1274]
From [here](https://github.com/elan-ev/tobira/pull/1273#issuecomment-2460338979):

> > Another comment regarding the date added: When and why did webpages start to display the "... ago" information instead of a date? I mean, there's a calendar symbol right there, so why not use an actual date format? Also, this seems to lead to misinterpretations unless 2010 is 15 years ago:
>
> That particular video is from 2010-01-22, so it's 14 years and 9.5 months ago, which gets rounded up to 15. 
>
> But re main point: for one, you can hover over the "15 years ago" to see the exact date. But secondly, I get your point, I am also sometimes annoyed by these relative dates. They can be handy as well though, especially when you have lots of "somewhat recent" videos. We can certainly change this though, for example everything that's longer than 6 months ago is shown as explicit date? :thinking: 

So yes... what's everyone's preference here? 

I think we shouldn't always show absolute dates: If it's less than a day ago, we would need to show date and time, which means lots of numbers. I think "X hours ago" is way better. Even if the difference is a few days, I think I prefer "X days ago". Otherwise I have to remember the date of today. Beyond that I could be convinced that the absolute day is better. Once the difference is larger than a few years, I am fairly sure the date is more useful than the relative time.

**Comment by dagraf** [comment]
I have nothig against the acutal solution ("x hours ago" or "x days ago" with specific date showing up when hovering over) and would suggest to leave it as it is and to invest our time elsewhere.

**Comment by LukasKalbertodt** [comment]
> invest our time elsewhere.

Implementation-wise this change is trivial, we just have to decide the threshold. But sure, we still invest time in discussing this.

**Comment by oas777** [comment]
PS: @sc-lmu, I didn't wait for you here because there's already a majority to ignore this for the time being. Feel free to chime in anyway.

**Comment by LukasKalbertodt** [comment]
@oas777 [wrote](https://github.com/elan-ev/tobira/issues/1395#issue-2988422791):

> We had complaints about the videos not being easily relatable when the date is in the "... ago" form. Therefore, this format should either be customizable or switched to the "DD.MM.YYYY" / "YYYY/MM/DD" format.
>
> ![Image](https://github.com/user-attachments/assets/a3917348-cb2e-410f-88c8-e4e03662561e)

**Comment by LukasKalbertodt** [comment]
I returned to this issue since it came up again in #1395 and #1313. I read a bit about this topic now, but while there is lots of discussions about it in UX-circles, there are lots of bad arguments flying around and it doesn't seem like there is any hard data to rely on. Let me try to summarize the problem.

---

First important point: we are dealing with two pretty independent decisions: "rounding/omitting information" and "relative vs absolute". All four combinations are possible: "15 years ago", "2010", "3 hours 17min 29sec ago" and "14th April 2025, 11:41:38".

### Rounding/omitting information

Should we show the exact datetime up to seconds or omit some information? I have two fairly strong opinions here:
- In almost all situations, omitting _some_ information is useful to reduce visual noise and speed up user comprehension. If something is months ago, hardly anyone will care about the seconds. If something is minutes ago, showing the full date is often not useful. And visual noise (lots of numbers) is a real problem, from design _and_ UX perspective. Of course, the amount of rounding/omission we need to discuss.
- There should always be an easy way to get the exact datetime as a user, e.g. by hovering over the date.

### Absolute vs relative

Humans think about time and dates in various ways. When setting a timer for tea, we don't care about the absolute time. For lecture recordings, we might be used to thinking in 2h time slots like 8 oclock, 10 oclock, etc. Oftentimes, we think in terms of week days instead of the date, certainly for short term plans, like picnic on Friday. When planning around Christmas, suddenly planning in dates (visiting you on the 25th) becomes natural. For a lecture that takes place once a week, one might only think about it in terms of "this week's recording vs last week's recording". Going back further, thinking about semesters is typical. This is just a small sample of how thinking patterns can differ depending on the situation.

Both relative and absolute datetimes have advantages and disadvantages, depending on the situation.

---

To answer both questions, we need to really consider the use case and try to understand how our users think. 

The following is how I as a user would think about datetimes (I'm not necessarily saying Tobira should do it exactly like that):
- For lecture recordings, looking at just one series/lecture:
  - Current semester: "Thursday", "Tuesday", "Thursday, a week ago", ..., "Tuesday, 3 weeks ago"
  - Past semesters: mhhh hard to say, maybe "WS 2023, Tuesday, KW 47" (WS being winter semester)? Or "WS 2023, week 3, Tuesday" with "week" just counting up for the semester? 
  - Point being: I certainly don't care about the date and would always have to convert the date to "day of week" 
- For lecture recordings when mixing many lectures, e.g. in "my bookmarks" that contain all lectures I'm attending
  - Current semester: "Monday 14:00", "Tuesday, 8:00", "Thursday, 16:00, a week ago", ...
  - Past semesters: see above, plus time
- For other, non-regular videos (e.g. promotional university videos, student explainer videos, ...), and for mostly everything else
  - Very recent: "3 min ago". Absolute times don't work well, as I'm rarely aware of the current time to that kind of precision.
  - Still today: "5 hours ago" or "11:25" would both work for me.
  - Days ago: either "2 days ago" or "Tuesday"
  - Weeks ago: "3 weeks ago" or absolute date (without time, maybe without year?)
  - Longer ago: relative time makes little sense I think, as I'm always aware of the current year and month, so seeing a date, I can easily grasp roughly how many months ago it is. Maybe omit the year if it's the current year?

Again, I don't think we should implement this exact logic as (a) other people's thinking pattern likely differ from mine, and (b) it would actually be quite complicated implementation wise (e.g. how to distinguish lecture recordings from others?). What I noticed, writing this down, is how often I think about "days of the week". I guess it makes sense, since a week is one of the main time frameworks we live in in our society. On the other hand, "day of month" (which is what most date formats use) is pretty irrelevant. I'm also usually unaware of today's "day of month", i.e. date.

I'm interested in how others think about all of this and where the thinking pattern differs. 

And I'm sorry that this comment is an in-depth "lets open this box and properly discuss it", rather than a "here is a specific solution". But I'm sure we can agree on something specific soon enough.





We do have two maybe slightly different places for dates: the "public" one that's next to videos on the homepage, in the search, etc; and the one in "My Videos" and "My Series". Maybe the use cases are different enough to warrant slightly different approaches, but I still hope we can use the exact same logic for both.

**Comment by oas777** [comment]
Not the comment I expected ... obviously, mine is more towards the "here is a specific solution" side of things: 

- 80% of our content is lecture recording, it will be similar or more at other universities, so maybe focus on that use case? Is someone can afford the customization I suggested, please come forward.
- I received nine replies regarding the pilot at ETH so far, three of them asked for the date to be shown, so maybe focus on that use case?
- While I agree the name of the day might be useful to distinguish Monday's recording from Wednesday's, the limitations of the UI will allow us to present only one information (date, time, name of day) and that would be the date for me.
- Time is the least important because students are already in the process of searching in a specific department and semester when they come across the series overview and there won't be two lectures in one day you differentiate by time.

TLDR: Get rid of "... ago" and give us a date.

**Comment by oas777** [comment]
Be honest: Did someone remember the hover-over?

![Image](https://github.com/user-attachments/assets/22f813fb-8078-4a1d-9f1b-29431764bd24)

**Comment by dagraf** [comment]
Looking at @LukasKalbertodt first comment here (https://github.com/elan-ev/tobira/issues/1274#issue-2638905090) what about a compromise? Maybe thats my Swiss blood forcing me to look for compromises...:
- Video with same date as actual date (meaning all videos from 'today'): "x hours ago"
- Video from the day before: "Yesterday"
- All older videos: specific date (without time)

**Comment by dagraf** [comment]
Good to know: Other video portal with a focus on lecture captures show specific dates (SWITCHtube, UZH-Mediaspace). SWITCHtube has "Today at hh:mm" for recordings from today.

<img width="1233" alt="Image" src="https://github.com/user-attachments/assets/198fa2e4-7b0d-476f-a85f-d6f818d0dcff" />

**Comment by oas777** [comment]
I presume Lukas knows what's coming: What time exactly? Recording starts at 10:13 am here at ETH, lecture starts at 10:15 am officially, with a lecturer from Bern running late that would be 10:16 am when the video is edited accordingly. Don't let anyone add a video to the lecture series for a missing recording because that might be 10:22 - pm!

Jokes aside, the compromise could be
- Video with same date as actual date (meaning all videos from 'today'): "Today" - or the actual date, no time.
- Video from the day before: "Yesterday" - or the actual date, no time.
- All older videos: specific date (without time)

**Comment by LukasKalbertodt** [comment]
I finally had time to properly think about this and re-read your comments. I suggest the following:

| Age        | de              | en                  | Alternatives/open questions
| ---------- | --------------- | ------------------- | ---------
| <= 60min   | Vor 38min       | 38min ago           |           
| Same day   | Heute, 14:00    | Today, 2:00 pm      | Or "Today at/um ..."               
| Yesterday  | Gestern, 8:13   | Yesterday, 8:13 am  | Or "Yesterday at/um ..."                   
| Same week  | Dienstag, 15:48 | Tuesday, 3:48 pm    | Omit time?                 
| Same year  | Mo, 5. April    | Mon, April 5th      | Include time?               
| older      | 27.08.2019      | 27/08/2019          | Or ISO YYYY-MM-DD? Include day of week?

I think I slightly prefer the ISO YYYY-MM-DD format over "local" date formats. While it's always been used by technical people, one might assume that it's not a format intended for or used by normal people. I beg to differ! I think it's perfectly reasonable and is _more_ clear than any other format. In particular, this sidesteps the age old issue of GB DD/MM/YYY vs US MM/DD/YYYY. The German local format is pretty unambiguous luckily. But yes, in internal contexts, the ISO YYYY-MM-DD is actually common, and [DIN 5008, which are the de-facto German rules in many areas, also uses the ISO date format](https://de.wikipedia.org/wiki/DIN_5008#Schreibweise_des_Datums).

Let me know what you think of this suggestion! I'd love to have a decision by the end of our next meeting on Monday.

**Comment by oas777** [comment]
I wonder how someone praising ISO comes forward with such a ramified solution to an easy problem, but here we are. Answering your questions, thus following your path unwillingly:

- Today/Yesterday **at**,
- omit time (especially because 8:13 makes me wonder whether that's the starting time of the recording no one should see), and
- ISO for the win. My favorite anyway.

**Comment by LukasKalbertodt** [comment]
We discussed this just now. Protocol:

- All three (Olaf, David, Simone) preferred fewer different cases/formats/rows in the table
  - Remove the first case (<60min) and just use the same format as the line below (today at...)
  - Also remove "Same week" case, using the format of "same year"
  - "Today" and "tomorro" cases are fine
- All three also wanted to always show the year, arguing that it makes comparisons easier between videos and that omitting the year is too implicit to communicate "it's this year"
- Regarding the date format for the "older" case, we went back and forth a bunch.
  - For German, most preferred not ISO, but ISO would be fine as well
  - For English, David preferred the GB DD/MM/YYYY. The two other options are ISO and "8th Aug, 2019", both of which avoid the confusion between US and GB formats. 
  - If English would use "8th Aug, 2019", then German should use the same format
  - I don't think we really settled for something in the end.

Our conclusion was that I would make a new suggestion and talk to our designer about this. I will therefore write another comment here shortly.

---

> to an easy problem

Quick remark about that: this might seem like an easy problem and yes "just use ISO date+time" is technically a solution, but not a great one. In programming like this, there are a vast number of things that first seem trivial or simple to solve, when in reality they are very complex and/or have very subtle details to consider. And since "datetimes" are shown a lot in the UI and are looked at a lot, I think this is an important topic for Tobira and spending time discussing this is not wasted, even if it feels like it.

**Comment by LukasKalbertodt** [comment]
I talked to our designer today and we have a new suggestion now. It reduces the number of cases and resolves some of our open questions/discussions. You will notice however, that it's not quite "just ISO" and still contains different cases/formats. Her reasoning (which I agree with) is as follows:

- Having multiple cases/formats first seems like it would be confusing/harder to understand, but that's not actually true. Something like "38 min ago" is immediately obvious to everyone, even if most other videos show a date + time, for example. When reading, humans don't need to iterate through all possible formats and check which one fits -> it is just read and understood, as long as the format itself is easy to understand. 
- Special formats like "38 min ago" even improve comprehension, as humans understand that the video is very recent way faster, compared to first comparing the time (and god forbid, even a date) with the current time/date. 
- Most importantly: users are used to that from other services on the internet. Most services use relative dates for recent events, with the threshold usually being weeks or months. Some, like YouTube, even exclusively use relative dates. And basically all websites use different formats depending on how long ago the event is. So not only are users not confused by this, they almost expect it. In particular the case of a very recent video: if Tobira would show date + time, most users would intuitively assume it's already a bit older, since almost all other services would show a relative time instead.

She agreed with you in that we should not omit the year for "same year, but many months ago". She also said using "8 Aug 2019" date format for all languages is very likely the best as it's best for comprehension and has no ambiguity. She personally also mainly uses ISO, but still slightly preferred "8 Aug 2019" for Tobira.

With that, our suggestion is as such:

| Age        | Format              | Notes
| ---------- | ------------------- | ---------
| < 60min    | 38min ago           |           
| Same week  | $day, 1:20 pm       | Where $day is "today", "yesterday" or a weekday 
| older      | 5 Aug 2021          | For videos younger than 6 months, also show weekday


There is one minor open question: "Today, 8:13" or "Today at 8:13"? But the main one: are these formatting rules acceptable to everyone, especially considering the arguments above?

**Comment by oas777** [comment]
I would love to simply give this a thumbs up, but
- will "1:20 pm" be "13:30 Uhr" in the German version
- like "8. August 2021" for "5 Aug 2021"?
- I prefer "Today at" / "Heute um" and
- I definitively have to know what metadata field the time stems from.

**Comment by oas777** [comment]
> > like "8. August 2021" for "5 Aug 2021"?
> 
> I don't think Germany is in time zone that's 3 days ahead of Great Britain. Jokes aside, the German version would be "5. Aug 2021", i.e. with added point. Or would you prefer the months to be fully written out? The might take a lot of Space for Dezember and other months.

I prefer them fully written out, Luk. 

> > I definitively have to know what metadata field the time stems from.
> 
> This is an orthogonal discussion because we use different metadata fields in different situations. _But_ most of the time, what we are talking about here, it uses `startTime` if that is set, and `created` otherwise.

Pardon my ignorance, but what would these say for https://tobira.opencast.org/lectures/d-arch/2021/spring/051-0126-00L/v/FhBYQ8gYM88 e. g.?

**Comment by LukasKalbertodt** [comment]
> > > like "8. August 2021" for "5 Aug 2021"?
> > 
> > 
> > I don't think Germany is in time zone that's 3 days ahead of Great Britain. Jokes aside, the German version would be "5. Aug 2021", i.e. with added point. Or would you prefer the months to be fully written out? The might take a lot of Space for Dezember and other months.
> 
> I prefer them fully written out, Luk.

I see, but again, might be a bit crammed in the UI, Olaf Andreas Schulte. Lets see what the others say

> 
> > > I definitively have to know what metadata field the time stems from.
> > 
> > 
> > This is an orthogonal discussion because we use different metadata fields in different situations. _But_ most of the time, what we are talking about here, it uses `startTime` if that is set, and `created` otherwise.
> 
> Pardon my ignorance, but what would these say for https://tobira.opencast.org/lectures/d-arch/2021/spring/051-0126-00L/v/FhBYQ8gYM88 e. g.?

That particular event has `2019-05-02T10:43:00Z` as both `created` and `startTime` (most events have these two fields with the same value). With the suggestion above it would be shown as "2 May 2019" (en) or "2. Mai 2019" (de).

**Comment by dagraf** [comment]
I agree to Lukas' suggestion from this comment (https://github.com/elan-ev/tobira/issues/1274#issuecomment-3004093634).

- I prefer "Today at" / "Heute um", too.
- I slightly prefer the month to be written out.
